core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/AppCDSBuildStep.java

```diff
@@ -0,0 +1,211 @@
+    }
+
+    /**
+     * Waits for the application terminate by looking for the proper logs, or just giving up after some time
+     */
+    private void waitForApplicationToStart(Process process) throws IOException {
+        if (!process.isAlive()) {
+            log.debug("The process that was supposed to create 'classes.lst' exited.");
+            return;
+        }
+        // don't wait for more than 10 seconds
+        // TODO: make it configurable?
+        final int maxWaits = 200;
+        int waits = 0;
+        BufferedReader in = new BufferedReader(new InputStreamReader(process.getInputStream(), StandardCharsets.UTF_8));
+        while (true) {
+            if (!process.isAlive()) {
+                log.debug("The process that was supposed to create classes.lst exited");
+                break;
+            }
+            String line = in.readLine();
+            if (line != null) {
+                if (line.contains(Timing.INSTALLED_FEATURES) || line.contains(MainClassBuildStep.FAILED_TO_START_MESSAGE)) {
+                    process.destroy();
+                    break;
+                }
+            }
+            try {
+                Thread.sleep(50);
+            } catch (InterruptedException ignored) {
+            }
+
+            // we don't want to wait forever
+            if (++waits >= maxWaits) {
+                break;
+            }
+        }
+        try {
+            in.close();
+        } catch (IOException ignored) {
+        }
+    }
+
+    /**
+     * @return The path of the created app-cds.jsa file or null of the file was not created
+     */
+    private Path createAppCDS(JarBuildItem jarResult, String javaBin,
+            Path classesLstPath, boolean isFastFar) {
+
+        Path workingDirectory = isFastFar ? jarResult.getPath().getParent().getParent() : jarResult.getPath().getParent();
+        Path appCDSPath = workingDirectory.resolve("app-cds.jsa");
+
+        List<String> command = new ArrayList<>(6);
+        command.add(javaBin);
+        command.add("-Xshare:dump");
+        command.add("-XX:SharedClassListFile=" + classesLstPath.toAbsolutePath().toString());
+        // We use the relative paths because at runtime 'java -XX:SharedArchiveFile=... -jar ...' expects the AppCDS and jar files
+        // to match exactly what was used at build time.
+        // For that reason we also run the creation process from inside the output directory,
+        // The end result is that users can simply use 'java -XX:SharedArchiveFile=app-cds.jsa -jar app.jar'
+        command.add("-XX:SharedArchiveFile=" + appCDSPath.getFileName().toString());
+        command.add("--class-path");
+        command.add(jarResult.getPath().getFileName().toString());
+
+        if (log.isDebugEnabled()) {
+            log.debugf("Launching command: '%s' to create final AppCDS", String.join(" ", command));
+        }
+
+        int exitCode;
+        try {
+            exitCode = new ProcessBuilder(command)
```

Comment:
I think this approach is going to fail in a lot of cases, as you need all the production services to be present.

Reply:
It could certainly fail. I'm not saying it's a good solution - because it isn't, but even like this you still get a list classes :)